use.1
.TH use 1 "July 21, 2014"
.SH NAME
use
.SH SYNOPSIS
.B use [options]
.SH DESCRIPTION
.B use
is a tool that allows a user to modify a
shell's environment variables in accordance to defined products.
It currently modifies DYLD_LIBRARY_PATH, LD_LIBRARY_PATH, MANPATH, PATH, XFILESEARCHPATH, and any other
variable that is defined in the use.table file.
.SH OPTIONS
.TP
.BI -a
Display information about use.
.TP
.BI -d " product"
This option detaches the given product. That is, the environment
variables of the current shell will not reference the paths defined in the
use.table file for the given product. Please note that products are detached in a last-in-first-out order.
.TP
.BI -l
List all known products stored in use.table.
.TP
.BI -n " variable"
Prevent use from modifying the environment variable variable. Current supported variables
are DYLD_LIBRARY_PATH, LD_LIBRARY_PATH, MANPATH, PATH, and XFILESEARCHPATH.
.TP
.BI -p
Pretend mode, that is, show potential changes. This option supersedes the -q option.
.TP
.BI -q
Quiet mode. Also, all errors and warnings will be suppressed. This option does not interfere with the -p option.
.TP
.BI -s " type"
Set the shell type. Valid shells are: BASH, CSH, KSH, SH, and TCSH.
This option is not necessary as the use alias, if configured properly,
passes the shell type to the use.bin executable.
.TP
.BI -t " table"
Use the table table instead of the official use.table table.
.TP
.BI -u " product"
This option uses the given product. The current shell's environment variables
are modified to reference the paths defined in use.table for the given product. Please note that products are applied in a last-in-first-out order.
.SH FILES
.B
 /usr/local/bin/use.bash
.B
 /usr/local/bin/use.bin
.B
 /usr/local/bin/use.csh
.B
 /usr/local/bin/use.ksh
.B
 /usr/local/bin/use.sh
.B
 /usr/local/bin/use.tcsh
.B
 /usr/local/share/use.table
.SH NOTES
use will issue a warning for a product that is not located in the use.table file. It will also issue a warning if a product references a nonexistent path.
You will need to create a use alias for each shell that you intend to use. Please see the
ALIAS EXAMPLE(S) section for more information.
Please note that use will use the /tmp directory if it's unable to locate a user's home directory. Although use uses mkstemp(), the use of /tmp may raise security concerns.
.SH ALIAS EXAMPLE(S)
.IP "bash:"
alias use='. /usr/local/bin/use.bash'
.SH AUTHOR(S)
.B Alexis Megas


RELEASE-NOTES
1.12.1 (04/07/2015)

1. Removed casts from malloc().

1.12.0 (03/22/2015)

1. Cleaning.
2. The function updateVariable() should produce an error
   if _stdout_ or flags is not defined.

1.11.9 (07/23/2014)

1. Temporary files must be opened for reading.

1.11.8 (07/21/2014)

1. Do not print /dev/null to standard output if fopen() fails.
   If use is executed with elevated privileges, /dev/null may be removed.
2. Open temporary files for appending instead of appending and reading.
3. Use mkstemp() to create a temporary file.

1.11.7 (07/16/2014)

1. The user's home will now be used to house temporary files. If a home
   is not available, /tmp will be used.

1.11.6 (06/18/2014)

1. Modified the manual.
2. Version remains at 1.11.6.

1.11.6 (06/15/2014)

1. Use %ld specifier for pid_t and uid_t in snprintf().

1.11.5 (05/10/2014)

1. Added the g flag to debug builds.
2. Changed the default manual location to /usr/share/man/man1 in configure.pl.
3. Improved error messages.
4. Modified the manual page, informing the user that product order is
   important.

1.11.4 (05/03/2014)

1. Added support for DYLD_LIBRARY_PATH on OS X.
2. Added the t command-line option, allowing for alternate use tables.

1.11.3 (04/21/2014)

1. Compiler flags.
2. Environment variables that are not defined are not properly initialized.
   Corrected.

1.11.2 (04/18/2014)

1. Added compiler flags.
2. Version remains at 1.11.2.

1.11.2 (04/09/2014)

1. Shell scripts should not exit as terminals shall be closed.

1.11.1 (04/04/2014)

1. Clear buffers before and after fgets().
2. Test argv in validate().
3. Verify that stdout is valid.

1.11.0 (03/20/2014)

1. Inspect argv[].

1.10.9 (03/14/2014)

1. Guard against invalid _stdout_.

1.10.8 (02/28/2014)

1. Added the Wconversion and Wformat compiler flags.
2. Replaced GCC with CC in Makefile.mk.
3. Replaced GCC_OPTIONS with CC_OPTIONS in Makefile.mk.

1.10.7 (02/24/2014)

1. Added formats to all fprintf() statements.
2. Included codes in shell exits.
3. Use /bin instead of /usr/bin in shell scripts.

1.10.6 (02/18/2014)

1. Clear buffers before using snprintf().

1.10.5 (02/06/2014)

1.Guard against empty pointers.

1.10.4 (02/04/2014)

1. Corrected arguments to snprintf().
2. Protect against negative size_t values.

1.10.4 (02/01/2014)

1. Protect array boundaries.

1.10.2 (09/12/2013)

1. Improved parameter checking for strncpy().
2. Inspect pointers.
3. Replaced source with dot for BASH.
4. Replaced strcat() with strncat().

1.10.1 (09/03/2013)

1. Added PIE and stack-smashing protection.
2. Use strncpy() instead of strcpy().

1.10.0 (03/13/2009)

1. Modified the make files.
2. Version remains at 1.10.0.

1.10.0 (02/29/2008)

1. Resolved several compilation issues.

1.09.0 (11/02/2007)

1. Modified the manual page.
2. Version remains at 1.09.0.

1.09.0 (03/06/2007)

1. Minor flaw fixed with the new 1.09.0 addition. Also updated the manual
   page.

1.09.0 (02/28/2007)

1. Warn the user if a product references a nonexistent path.

1.08.0 (11/17/2006)

1. A new option (-p) has been introduced. Please read the man page for
   more information.

1.07.0 (10/24/2006)

1. A minor fix.

1.06.0 (09/22/2006)

1. This version allows variables to contain multiple paths, for example,
   ddd.PATH:/usr/local/bin:/opt/sfw/bin.

1.05.0 (09/17/2006)

1. Spelling corrections.
2. Version remains at 1.05.0.

1.05.0 (08/13/2006)

1. Detached products should be processed first.
2. Version remains at 1.05.0.

1.05.0 (08/02/2006)

1. Bug fix. The environment variable XFILESEARCHPATH was not being set
   properly.

1.04.0 (07/23/2006)

1. Bug fix. Use set crucial paths (MANPATH, PATH, etc.) even if they
   were not specified in the use.table file.

1.03.0 (07/23/2006)

1. A NULL envvar value caused allocenv() to misbehave.

1.02.0 (07/21/2006)

1. Updated use.tcsh.
2. Version remains at 1.02.0.

1.02.0 (07/19/2006)

1. Resolved an issue with the quiet option.

1.01.0 (07/18/2006)

1. This release includes some fixes to the shell scripts as well as
   the binary. A new Perl installation script has also been added. Please
   read the INSTALL file for complete information.

1.00.0 (07/16/2006)

1. Initial release.


use.c
/*
** Copyright (c) 2006 - present, Alexis Megas.
** All rights reserved.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions
** are met:
** 1. Redistributions of source code must retain the above copyright
**    notice, this list of conditions and the following disclaimer.
** 2. Redistributions in binary form must reproduce the above copyright
**    notice, this list of conditions and the following disclaimer in the
**    documentation and/or other materials provided with the distribution.
** 3. The name of the author may not be used to endorse or promote products
**    derived from this software without specific prior written permission.
**
** THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
** IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
** OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
** IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
** INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
** NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
** THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "use.h"
#include "validate.h"

static char *LD_LIBRARY_PATH = 0;
static char *MANPATH = 0;
static char *PATH = 0;
static char *XFILESEARCHPATH = 0;

static int allocenv(char **envvar,
		    const char *value,
		    const int action,
		    const struct flags_struct *flags);
static int prepare(FILE *fp,
		   const char *product,
		   const struct flags_struct *flags,
		   const int action);
static void reset(struct flags_struct *flags);
static int updatevariable(const char *variable,
			  const char *value,
			  const struct flags_struct *flags,
			  const int action);
static int use(struct flags_struct *flags);

int main(int argc, char *argv[])
{
  FILE *fp = 0;
  char filename[PATH_MAX];
  char line[MAX_LINE_LENGTH];
  char *tmp = 0;
  int fd = -1;
  int rc = 0;
  struct flags_struct flags;
  struct passwd *pw = getpwuid(getuid());

  (void) memset(TEMPDIR, 0, sizeof(TEMPDIR));

  if(pw && pw->pw_dir)
    (void) snprintf(TEMPDIR, sizeof(TEMPDIR), "%s", pw->pw_dir);
  else
    (void) snprintf(TEMPDIR, sizeof(TEMPDIR), "%s", "/tmp");

  (void) memset(filename, 0, sizeof(filename));
  (void) snprintf
    (filename, sizeof(filename), "%s/.use.sourceme.%ld.%ldXXXXXX",
     TEMPDIR, (long) getuid(), (long) getpid());

  if((fd = mkstemp(filename)) == -1)
    {
      if(stdout)
	fprintf(stdout, "%s", "");

      return EXIT_FAILURE;
    }

  if(!(_stdout_ = fdopen(fd, "a+")))
    {
      if(stdout)
	(void) fprintf(stdout, "%s", "");

      return EXIT_FAILURE;
    }
  else if(stdout)
    (void) fprintf(stdout, "%s", filename);

  (void) setvbuf(_stdout_, 0, _IONBF, (size_t) 0);
  _usefp_ = 0;

  /*
  ** Reset the flags structure.
  */

  reset(&flags);

  if(validate(argc, argv, &flags) != 0)
    {
      /*
      ** Print the usage information.
      */

      rc = 1;

      if(!flags.quiet)
	(void) fprintf(_stdout_, "%s", "echo \"use: incorrect usage.\"\n");

      goto done_label;
    }

  if(flags.list)
    {
      if(_usefp_ == 0 && !(fp = fopen(USETABLE, "r")))
	{
	  rc = 1;
	  (void) fprintf(_stdout_, "echo \"Error: unable to open %s.\"",
			 USETABLE);
	}
      else
	{
	  if(_usefp_)
	    fp = _usefp_;

	  (void) memset(line, 0, sizeof(line));

	  while(fgets(line, (int) sizeof(line), fp))
	    {
	      if(strstr(line, ".description") &&
		 strstr(line, ":") && line[0] != '#')
		{
		  tmp = strstr(line, ":") + 1;

		  if(tmp && strlen(tmp) > 0 && tmp[strlen(tmp) - 1] == '\n')
		    tmp[strlen(tmp) - 1] = 0;

		  if(tmp)
		    (void) fprintf(_stdout_, "echo \"%s\"\n", tmp);
		}

	      (void) memset(line, 0, sizeof(line));
	    }

	  (void) fclose(fp);
	  fp = 0;
	}

      goto done_label;
    }
  else if(flags.about)
    {
      (void) fprintf(_stdout_, "%s", "echo \"use\"\n");
      (void) fprintf(_stdout_, "echo \"Version: %s\"\n", VERSION);
      (void) fprintf(_stdout_, "echo \"Compilation Date & Time: %s %s\"\n",
		     __DATE__, __TIME__);
      (void) fprintf(_stdout_, "echo \"Temporary Directory: %s\"\n", TEMPDIR);
      (void) fprintf(_stdout_, "echo \"Use Table: %s\"\n", USETABLE);
      goto done_label;
    }

  rc = use(&flags);
  free(LD_LIBRARY_PATH);
  free(MANPATH);
  free(PATH);
  free(XFILESEARCHPATH);

 done_label:
  if(_stdout_)
    {
      (void) fflush(_stdout_);

      /*
      ** Also closes fd.
      */

      (void) fclose(_stdout_);
    }

  if(rc != 0)
    return EXIT_FAILURE;
  else
    return EXIT_SUCCESS;
}

static void reset(struct flags_struct *flags)
{
  int i = 0;

  if(flags)
    {
      flags->about = 0;
      flags->items_detached = 0;
      flags->items_used = 0;
      flags->list = 0;
      flags->no_ld_library_path = 0;
      flags->no_manpath = 0;
      flags->no_path = 0;
      flags->no_xfilesearchpath = 0;
      flags->pretend = 0;
      flags->quiet = 0;
      flags->shell_type = 0;

      for(i = 0; i < MAX_PRODUCTS; i++)
	{
	  (void) memset(flags->detached[i], 0, sizeof(flags->detached[0]));
	  (void) memset(flags->used[i], 0, sizeof(flags->used[0]));
	}
    }
}

static int use(struct flags_struct *flags)
{
  FILE *fp = 0;
  char envact[MAX_LINE_LENGTH];
  char line[MAX_LINE_LENGTH];
  char *product = 0;
  char *stdvalues[4]; /* This must be the same size as stdvariables. */
  char *stdvariables[] = 
    {
#if defined(__APPLE__) || defined(__MACH__)
      "DYLD_LIBRARY_PATH",
#else
      "LD_LIBRARY_PATH",
#endif
      "MANPATH",
      "PATH",
      "XFILESEARCHPATH"
    };
  char *tmp = 0;
  int dousevalues[4]; /* This must be the same size as stdvariables. */
  int i = 0;
  int rc = 0;
  size_t size = 0;

  if(!_stdout_ || !flags)
    {
      rc = 1;
      goto done_label;
    }

  if(flags->no_ld_library_path == 0)
    {
#if defined(__APPLE__) || defined(__MACH__)
      if((tmp = getenv("DYLD_LIBRARY_PATH")))
#else
      if((tmp = getenv("LD_LIBRARY_PATH")))
#endif
	size = strlen(tmp) + 1;
      else
	size = 0;

      free(LD_LIBRARY_PATH);
      LD_LIBRARY_PATH = 0;

      if(size > 0)
	{
	  LD_LIBRARY_PATH = malloc(size);
	  (void) memset(LD_LIBRARY_PATH, 0, size);
	}

      if(LD_LIBRARY_PATH && tmp)
	(void) strncpy(LD_LIBRARY_PATH, tmp, size - 1);
      else if(size > 0)
	{
	  rc = 1;

	  if(!flags->quiet)
#if defined(__APPLE__) || defined(__MACH__)
	    (void) fprintf(_stdout_, "%s",
			   "echo \"Error: unable to allocate memory for "
			   "DYLD_LIBRARY_PATH or tmp is empty.\"\n");
#else
	    (void) fprintf(_stdout_, "%s",
			   "echo \"Error: unable to allocate memory for "
			   "LD_LIBRARY_PATH or tmp is empty.\"\n");
#endif

	  goto done_label;
	}
    }
  else if(!flags->quiet)
#if defined(__APPLE__) || defined(__MACH__)
    (void) fprintf(_stdout_, "%s",
		   "echo \"Warning: DYLD_LIBRARY_PATH not "
		   "updated.\"\n");
#else
    (void) fprintf(_stdout_, "%s",
		   "echo \"Warning: LD_LIBRARY_PATH not "
		   "updated.\"\n");
#endif

  if(flags->no_manpath == 0)
    {
      if((tmp = getenv("MANPATH")))
	size = strlen(tmp) + 1;
      else
	size = 0;

      free(MANPATH);
      MANPATH = 0;

      if(size > 0)
	{
	  MANPATH = malloc(size);
	  (void) memset(MANPATH, 0, size);
	}

      if(MANPATH && tmp)
	(void) strncpy(MANPATH, tmp, size - 1);
      else if(size > 0)
	{
	  rc = 1;

	  if(!flags->quiet)
	    (void) fprintf(_stdout_, "%s",
			   "echo \"Error: unable to allocate "
			   "memory for MANPATH or tmp is empty.\"\n");

	  goto done_label;
	}
    }
  else if(!flags->quiet)
    (void) fprintf(_stdout_, "%s",
		   "echo \"Warning: MANPATH not updated.\"\n");

  if(flags->no_path == 0)
    {
      if((tmp = getenv("PATH")))
	size = strlen(tmp) + 1;
      else
	size = 0;

      free(PATH);
      PATH = 0;

      if(size > 0)
	{
	  PATH = malloc(size);
	  (void) memset(PATH, 0, size);
	}

      if(PATH && tmp)
	(void) strncpy(PATH, tmp, size - 1);
      else
	{
	  rc = 1;

	  if(!flags->quiet)
	    (void) fprintf(_stdout_, "%s",
			   "echo \"Error: unable to allocate "
			   "memory for PATH or tmp is empty.\"\n");

	  goto done_label;
	}
    }
  else if(!flags->quiet)
    (void) fprintf(_stdout_, "%s", "echo \"Warning: PATH not updated.\"\n");

  if(flags->no_xfilesearchpath == 0)
    {
      if((tmp = getenv("XFILESEARCHPATH")))
	size = strlen(tmp) + 1;
      else
	size = 0;

      free(XFILESEARCHPATH);
      XFILESEARCHPATH = 0;

      if(size > 0)
	{
	  XFILESEARCHPATH = malloc(size);
	  (void) memset(XFILESEARCHPATH, 0, size);
	}

      if(XFILESEARCHPATH && tmp)
	(void) strncpy(XFILESEARCHPATH, tmp, size - 1);
      else if(size > 0)
	{
	  rc = 1;

	  if(!flags->quiet)
	    (void) fprintf(_stdout_, "%s",
			   "echo \"Error: unable to allocate memory for "
			   "XFILESEARCHPATH or tmp is empty.\"\n");

	  goto done_label;
	}
    }
  else if(!flags->quiet)
    (void) fprintf(_stdout_, "%s",
		   "echo \"Warning: XFILESEARCHPATH not "
		   "updated.\"\n");

  if(_usefp_)
    fp = _usefp_;
  else
    fp = fopen(USETABLE, "r");

  if(fp)
    {
      int found = 0;

      for(i = flags->items_detached - 1; i >= 0; i--)
	{
	  found = 0;
	  (void) memset(line, 0, sizeof(line));

	  while(fgets(line, (int) sizeof(line), fp))
	    {
	      if(strstr(line, ".description") &&
		 strstr(line, ":") && line[0] != '#')
		{
		  product = line;

		  if(strstr(product, ".") &&
		     strlen(product) - strlen(strstr(product, ".")) > 0)
		    {
		      product[strlen(product) -
			      strlen(strstr(product, "."))] = 0;

		      if(strcmp(flags->detached[i], product) == 0)
			{
			  found = 1;

			  if(prepare(fp, product, flags, DELETE_PATH) != 0)
			    {
			      rc = 1;

			      if(!flags->quiet)
				(void) fprintf
				  (_stdout_,
				   "echo \"Error: unable to prepare "
				   "containers for product %s. Please "
				   "review the use file.\"\n",
				   product);

			      goto done_label;
			    }
			  else if(!flags->quiet && !flags->pretend)
			    (void) fprintf(_stdout_, "echo \"%s has been "
					   "detached successfully.\"\n",
					   product);

			  break;
			}
		    }
		  else
		    {
		      rc = 1;

		      if(!flags->quiet)
			{
			  if(_usefp_)
			    (void) fprintf(_stdout_,
					   "%s",
					   "echo \"Error: possible "
					   "misconfiguration with "
					   "the provided table file.\"\n");
			  else
			    (void) fprintf(_stdout_,
					   "echo \"Error: possible "
					   "misconfiguration with %s.\"\n",
					   USETABLE);
			}

		      break;
		    }
		}

	      (void) memset(line, 0, sizeof(line));
	    }

	  if(found != 1 && !flags->quiet)
	    (void) fprintf(_stdout_, "echo \"Warning: product %s "
			   "not found.\"\n",
			   flags->detached[i]);

	  rewind(fp);
	}

      for(i = flags->items_used - 1; i >= 0; i--)
	{
	  found = 0;
	  (void) memset(line, 0, sizeof(line));

	  while(fgets(line, (int) sizeof(line), fp))
	    {
	      if(strstr(line, ".description") &&
		 strstr(line, ":") && line[0] != '#')
		{
		  product = line;

		  if(strstr(product, ".") &&
		     strlen(product) - strlen(strstr(product, ".")) > 0)
		    {
		      product[strlen(product) -
			      strlen(strstr(product, "."))] = 0;

		      if(strcmp(flags->used[i], product) == 0)
			{
			  found = 1;

			  if(prepare(fp, product, flags, ADD_PATH) != 0)
			    {
			      rc = 1;

			      if(!flags->quiet)
				(void) fprintf
				  (_stdout_,
				   "echo \"Error: unable to prepare "
				   "containers for product %s. Please "
				   "review the use file.\"\n",
				   product);

			      goto done_label;
			    }
			  else if(!flags->quiet && !flags->pretend)
			    (void) fprintf(_stdout_, "echo \"%s has been "
					   "accessed successfully.\"\n",
					   product);

			  break;
			}
		    }
		  else
		    {
		      rc = 1;

		      if(!flags->quiet)
			{
			  if(_usefp_)
			    (void) fprintf(_stdout_,
					   "%s",
					   "echo \"Error: possible "
					   "misconfiguration with "
					   "the provided table file.\"\n");
			  else
			    (void) fprintf(_stdout_,
					   "echo \"Error: possible "
					   "misconfiguration with %s.\"\n",
					   USETABLE);
			}

		      break;
		    }
		}

	      (void) memset(line, 0, sizeof(line));
	    }

	  if(found != 1 && !flags->quiet)
	    (void) fprintf(_stdout_, "echo \"Warning: product %s "
			   "not found.\"\n", flags->used[i]);

	  rewind(fp);
	}
    }
  else
    {
      rc = 1;

      if(!flags->quiet)
	{
	  if(_usefp_)
	    (void) fprintf(_stdout_, "%s",
			   "echo \"Error: unable to "
			   "open the provided table file.\"\n");
	  else
	    (void) fprintf(_stdout_, "echo \"Error: unable to "
			   "open %s.\"\n", USETABLE);
	}
    }

 done_label:

  if(fp)
    {
      (void) fclose(fp);
      fp = 0;
    }

  if(rc == 0)
    {
      /*
      ** Update the sourced file.
      */

      stdvalues[0] = LD_LIBRARY_PATH;
      stdvalues[1] = MANPATH;
      stdvalues[2] = PATH;
      stdvalues[3] = XFILESEARCHPATH;
      dousevalues[0] = flags->no_ld_library_path;
      dousevalues[1] = flags->no_manpath;
      dousevalues[2] = flags->no_path;
      dousevalues[3] = flags->no_xfilesearchpath;

      for(i = 0; i < 4; i++)
	if(stdvalues[i] != 0 &&
	   strlen(stdvalues[i]) > 0) /* In case the -n option was used. */
	  {
	    (void) memset(envact, 0, sizeof(envact));

	    if(flags->shell_type == BASH)
	      (void) snprintf(envact, sizeof(envact), "export %s=%s",
			      stdvariables[i],
			      stdvalues[i]);
	    else if(flags->shell_type == CSH)
	      (void) snprintf(envact, sizeof(envact), "setenv %s %s",
			      stdvariables[i],
			      stdvalues[i]);
	    else if(flags->shell_type == KSH)
	      (void) snprintf(envact, sizeof(envact), "export %s=%s",
			      stdvariables[i],
			      stdvalues[i]);
	    else if(flags->shell_type == SH)
	      (void) snprintf(envact, sizeof(envact), "export %s=%s",
			      stdvariables[i],
			      stdvalues[i]);
	    else if(flags->shell_type == TCSH)
	      (void) snprintf(envact, sizeof(envact), "setenv %s %s",
			      stdvariables[i],
			      stdvalues[i]);
	    else
	      continue;

	    if(flags->pretend)
	      (void) fprintf(_stdout_, "echo \"%s\"\n", envact);
	    else
	      (void) fprintf(_stdout_, "%s\n", envact);
	  }
	else if(dousevalues[i] == 0)
	  {
	    (void) memset(envact, 0, sizeof(envact));

	    if(flags->shell_type == BASH)
	      (void) snprintf(envact, sizeof(envact), "unset %s",
			      stdvariables[i]);
	    else if(flags->shell_type == CSH)
	      (void) snprintf(envact, sizeof(envact), "unsetenv %s",
			      stdvariables[i]);
	    else if(flags->shell_type == KSH)
	      (void) snprintf(envact, sizeof(envact), "unset %s",
			      stdvariables[i]);
	    else if(flags->shell_type == SH)
	      (void) snprintf(envact, sizeof(envact), "unset %s",
			      stdvariables[i]);
	    else if(flags->shell_type == TCSH)
	      (void) snprintf(envact, sizeof(envact), "unsetenv %s",
			      stdvariables[i]);
	    else
	      continue;

	    if(flags->pretend)
	      (void) fprintf(_stdout_, "echo \"%s\"\n", envact);
	    else
	      (void) fprintf(_stdout_, "%s\n", envact);
	  }
    }

  return rc;
}

static int prepare(FILE *fp, const char *product,
		   const struct flags_struct *flags,
		   const int action)
{
  char *lasts;
  char line[MAX_LINE_LENGTH];
  char *value = 0;
  char *variable = 0;
  int rc = 0;

  /*
  ** The order of the products listed is significant. Products at the
  ** front of the list have a higher priority than products at the end.
  */

  if(!flags || !fp || !product || strlen(product) == 0)
    {
      rc = 1;
      return rc;
    }

  (void) memset(line, 0, sizeof(line));

  while(fgets(line, (int) sizeof(line), fp))
    {
      if(strncmp(line, product, strlen(product)) == 0 && strstr(line, ":"))
	{
	  if(strtok_r(line, ".", &lasts))
	    variable = strtok_r(0, ":", &lasts);

	  if(!(value = strtok_r(0, ":", &lasts)))
	    value = "";

	  do
	    {
	      if(value && strlen(value) > 0 &&
		 value[strlen(value) - 1] == '\n')
		value[strlen(value) - 1] = 0;

#ifdef DEBUG
	      if(value && variable)
		(void) fprintf(stderr, "%s=%s\n", variable, value);
#endif

	      if(updatevariable(variable, value, flags, action) != 0)
		{
		  rc = 1;
		  break;
		}

	      value = strtok_r(0, ":", &lasts);
	    }
	  while(rc == 0 && value);

	  if(rc != 0)
	    break;
	}
      else
	break;

      (void) memset(line, 0, sizeof(line));
    }

  return rc;
}

static int updatevariable(const char *variable, const char *value,
			  const struct flags_struct *flags, const int action)
{
  char envact[MAX_LINE_LENGTH];
  int rc = 0;

  if(!_stdout_ || !flags || !value || !variable || strlen(variable) == 0)
    {
      if(!_stdout_ || !flags)
	rc = 1;
    }
#if defined(__APPLE__) || defined(__MACH__)
  else if(strcmp(variable, "DYLD_LIBRARY_PATH") == 0)
#else
  else if(strcmp(variable, "LD_LIBRARY_PATH") == 0)
#endif
    {
      if(flags->no_ld_library_path == 0)
	rc = allocenv(&LD_LIBRARY_PATH, value, action, flags);
    }
  else if(strcmp(variable, "MANPATH") == 0)
    {
      if(flags->no_manpath == 0)
	rc = allocenv(&MANPATH, value, action, flags);
    }
  else if(strcmp(variable, "PATH") == 0)
    rc = allocenv(&PATH, value, action, flags);
  else if(strcmp(variable, "XFILESEARCHPATH") == 0)
    {
      if(flags->no_xfilesearchpath == 0)
	rc = allocenv(&XFILESEARCHPATH, value, action, flags);
    }
  else
    {
      /*
      ** Update the sourced file.
      */

      (void) memset(envact, 0, sizeof(envact));

      if(action == ADD_PATH)
	{
	  if(validatePath(value, flags) != 0)
	    (void) fprintf(_stdout_, "echo \"Warning: %s "
			   "is not a valid path.\"\n", value);

	  if(flags->shell_type == BASH)
	    (void) snprintf(envact, sizeof(envact), "export %s=%s",
			    variable, value);
	  else if(flags->shell_type == CSH)
	    (void) snprintf(envact, sizeof(envact), "setenv %s %s",
			    variable, value);
	  else if(flags->shell_type == KSH)
	    (void) snprintf(envact, sizeof(envact), "export %s=%s",
			    variable, value);
	  else if(flags->shell_type == SH)
	    (void) snprintf(envact, sizeof(envact), "export %s=%s",
			    variable, value);
	  else if(flags->shell_type == TCSH)
	    (void) snprintf(envact, sizeof(envact), "setenv %s %s",
			    variable, value);
	}
      else
	{
	  if(flags->shell_type == BASH)
	    (void) snprintf(envact, sizeof(envact), "unset %s", variable);
	  else if(flags->shell_type == CSH)
	    (void) snprintf(envact, sizeof(envact), "unsetenv %s", variable);
	  else if(flags->shell_type == KSH)
	    (void) snprintf(envact, sizeof(envact), "unset %s", variable);
	  else if(flags->shell_type == SH)
	    (void) snprintf(envact, sizeof(envact), "unset %s", variable);
	  else if(flags->shell_type == TCSH)
	    (void) snprintf(envact, sizeof(envact), "unsetenv %s", variable);
	}

      if(strlen(envact) > 0)
	{
	  if(flags->pretend)
	    (void) fprintf(_stdout_, "echo \"%s\"\n", envact);
	  else
	    (void) fprintf(_stdout_, "%s\n", envact);
	}
    }

  return rc;
}

static int allocenv(char **envvar, const char *value, const int action,
		    const struct flags_struct *flags)
{
  char *tmp = 0;
  char *token = 0;
  int found = 0;
  int rc = 0;
  size_t size = 0;

  if(!_stdout_ || !flags)
    goto done_label;
  else if(action == ADD_PATH)
    {
      if(*envvar && value && strlen(value) > 0)
	size = strlen(*envvar) + strlen(value) + strlen(":") + 1;
      else if(value && strlen(value) > 0)
	size = strlen(value) + strlen(":") + 1;
    }
  else
    {
      if(*envvar)
	size = strlen(*envvar) + 1;
      else
	/*
	** Not an error.
	*/

	goto done_label;
    }

  if(size == 0)
    {
      rc = 1;
      goto done_label;
    }

  if(!(tmp = malloc(size)))
    {
      rc = 1;
      goto done_label;
    }

  (void) memset(tmp, 0, size);

  if(action == ADD_PATH && value && strlen(value) > 0)
    {
      if(validatePath(value, flags) != 0)
	(void) fprintf(_stdout_, "echo \"Warning: %s "
		       "is not a valid path.\"\n", value);

      if(size > strlen(tmp))
	(void) strncat(tmp, value, size - strlen(tmp) - 1);

      if(size > strlen(tmp))
	(void) strncat(tmp, ":", size - strlen(tmp) - 1);

#ifdef DEBUG
      (void) fprintf(stderr, "PATH1 = %s\n", value);
#endif
    }

  token = strtok(*envvar, ":");

  while(token != 0)
    {
      if(value && (found = strcmp(token, value)) != 0)
	if(size > strlen(tmp))
	  (void) strncat(tmp, token, size - strlen(tmp) - 1);

      if((token = strtok(0, ":")) && found != 0)
	if(size > strlen(tmp))
	  (void) strncat(tmp, ":", size - strlen(tmp) - 1);
    }

  if(strlen(tmp) > 0 && tmp[strlen(tmp) - 1] == ':')
    tmp[strlen(tmp) - 1] = 0;

  free(*envvar);
  size = strlen(tmp) + 1;

  if(!(*envvar = malloc(size)))
    {
      rc = 1;
      goto done_label;
    }
  else
    {
      (void) memset(*envvar, 0, size);
      (void) snprintf(*envvar, size, "%s", tmp);
    }

 done_label:
  free(tmp);
  return rc;
}
